src/graaflib/algorithm/minimum_spanning_tree.h

```diff
@@ -0,0 +1,75 @@
+#pragma once
+
+#include <graaflib/graph.h>
+#include <graaflib/types.h>
+
+#include <algorithm>
+#include <utility>
+#include <vector>
+
+namespace graaf::algorithm {
+
+// Helper struct for representing disjoint sets
+struct DisjointSet {
+  std::vector<vertex_id_t> parent;
+  std::vector<size_t> rank;
+
+  DisjointSet(size_t num_vertices) {
+    parent.resize(num_vertices);
+    rank.resize(num_vertices, 0);
+    for (vertex_id_t i = 0; i < num_vertices; ++i) parent[i] = i;
+  }
+
+  vertex_id_t find(vertex_id_t vertex) {
+    if (vertex != parent[vertex]) parent[vertex] = find(parent[vertex]);
+    return parent[vertex];
+  }
+
+  void union_sets(vertex_id_t vertex1, vertex_id_t vertex2) {
+    vertex_id_t root1 = find(vertex1);
+    vertex_id_t root2 = find(vertex2);
+    if (root1 != root2) {
+      if (rank[root1] < rank[root2]) std::swap(root1, root2);
+      parent[root2] = root1;
+      if (rank[root1] == rank[root2]) ++rank[root1];
+    }
+  }
+};
+
+template <typename V, typename E, graph_spec S>
+std::vector<std::pair<vertex_id_t, vertex_id_t>> minimum_spanning_tree(
+    const graph<V, E, S>& graph) {
+  using Edge = std::pair<edge_weight_t, std::pair<vertex_id_t, vertex_id_t>>;
```

Comment:
Same here, could you move the implementation to the `.tpp` file to keep you interface clean here? It would also be nice to have a short docstring on what this function does.

Maybe you could take a look at `algorithm/graph_traversal.h` for inspiration.